### PR TITLE
dm/compare-ws-grpc

### PR DESCRIPTION
**Pull request by daniellemachpud**

### CODE REVIEW

**Review by juliette-grantham — APPROVED** [review]

**Comment by juliette-grantham** [comment]
Do we want to add average time difference?

**Comment by daniellemachpud** [comment]
> Do we want to add average time difference?

yes, and we should also do it in micro / nano seconds, because ms is too high
also, in order to do the best compare, we want to include all fields in the include session, and right now in websocket we include only `tx_hash`. so need to make some changes  in `newSubTxFeedRequestBX` function, and allow to send all fields

File: go/cmd/ethcompare/main.go [hunk 1]
```diff
@@ -76,6 +76,31 @@ func main() {
 				},
 				Action: cmptxspeed.NewTxSpeedCompareService().Run,
 			},
+			{
+				Name:  "txwsgrpc",
+				Usage: "compares stream of txs from ws vs grpc",
```
Comment (Uri-Margalit-bloXroute) [inline go/cmd/ethcompare/main.go]:
```suggestion
				Usage: "compares stream of txs from gateway grpc vs gateway websocket",
```

File: go/internal/pkg/flags/flags.go [hunk 2]
```diff
@@ -7,7 +7,12 @@ var (
 	Gateway = &cli.StringFlag{
 		Name:  "gateway",
 		Usage: "gateway websocket connection string",
-		Value: "ws://127.0.0.1:28333/ws",
+		Value: "ws://127.0.0.1:28346/ws",
```
Comment (Uri-Margalit-bloXroute) [inline go/internal/pkg/flags/flags.go]: why change the default ?

File: go/pkg/cmpfeeds/compare_ws_grpc_tx_feeds.go [hunk 3]
```diff
@@ -0,0 +1,673 @@
...
+	"performance/internal/pkg/flags"
+	"performance/internal/pkg/utils"
+	"performance/internal/pkg/ws"
+	pb "performance/pkg/cmpfeeds/protobuf"
+	"strings"
+	"sync"
+	"time"
+)
+
+// TxWsGrpcCompareService represents a service which compares transaction feeds time difference
```
Comment (Uri-Margalit-bloXroute) [inline go/pkg/cmpfeeds/compare_ws_grpc_tx_feeds.go]: please change, the first should be the faster.
`TxGrpcWSCompareService`

File: go/pkg/cmpfeeds/compare_ws_grpc_tx_feeds.go [hunk 4]
```diff
@@ -0,0 +1,673 @@
...
+	if newTxSeenByBothFeeds != 0 {
+		txPercentageSeenByGatewayFirst = int(
+			(float64(txSeenByBothFeedsGatewayFirst) / float64(newTxSeenByBothFeeds)) * 100)
+	}
+
+	results := fmt.Sprintf(
+		"\nAnalysis of Transactions received on both feeds:\n"+
+			"Number of transactions: %d\n"+
+			"Number of transactions received from websocket connection first: %d\n"+
+			"Number of transactions received from gRPC connection first: %d\n"+
```
Comment (Uri-Margalit-bloXroute) [inline go/pkg/cmpfeeds/compare_ws_grpc_tx_feeds.go]: switch between the lines

File: go/pkg/cmpfeeds/compare_ws_grpc_tx_feeds.go [hunk 5]
```diff
@@ -0,0 +1,673 @@
...
+			"Percentage of transactions seen first from websocket connection: %d%%\n"+
```
Comment (Uri-Margalit-bloXroute) [inline go/pkg/cmpfeeds/compare_ws_grpc_tx_feeds.go]: show the data from the grpc point of view

File: go/pkg/cmpfeeds/compare_ws_grpc_tx_feeds.go [hunk 6]
```diff
@@ -0,0 +1,673 @@
...
+			"Average time difference for transactions received first from websocket connection (us): %d\n"+
+			"Average time difference for transactions received first from gRPC connection (us): %d\n"+
```
Comment (Uri-Margalit-bloXroute) [inline go/pkg/cmpfeeds/compare_ws_grpc_tx_feeds.go]: switch between the lines

File: go/pkg/cmpfeeds/compare_ws_grpc_tx_feeds.go [hunk 7]
```diff
@@ -0,0 +1,673 @@
...
+			"\nTotal Transactions summary:\n"+
+			"Total tx from ws connection: %d\n"+
+			"Total tx from gRPC connection: %d\n"+
```
Comment (Uri-Margalit-bloXroute) [inline go/pkg/cmpfeeds/compare_ws_grpc_tx_feeds.go]: switch between the lines

File: go/pkg/cmpfeeds/compare_ws_grpc_tx_feeds.go [hunk 8]
```diff
@@ -0,0 +1,673 @@
...
+		txReceivedByGatewayFirstAvgDelta,
+		txReceivedByGrpcGatewayFirstAvgDelta,
+		totalTxFromGateway,
+		totalTxFromGrpcGateway,
+		len(s.lowFeeHashes))
+
+	verboseResults := fmt.Sprintf(
+		"Number of high delta tx ignored: %d\n"+
+			"Number of new transactions received first from gateway: %d\n"+
+			"Number of new transactions received first from node: %d\n"+
```
Comment (Uri-Margalit-bloXroute) [inline go/pkg/cmpfeeds/compare_ws_grpc_tx_feeds.go]: correct strings
gateway -> grpc connection
node -> websocket connection

File: go/internal/pkg/flags/flags.go (hunk 2 above)
Comment (daniellemachpud) [reply go/internal/pkg/flags/flags.go]: because of my local gateway.. will change

**Review by Uri-Margalit-bloXroute — APPROVED** [review]

**Pull request merged**